File: packages/sdk/src/wallet.ts
/**
 * Phettagotchi Wallet — Auto-generate and manage a Solana keypair
 *
 * Stores keypair at ~/.phettagotchi/keypair.json
 * Used by the MCP server for automatic wallet provisioning.
 */

import { Keypair, Transaction, VersionedTransaction } from '@solana/web3.js';
import * as fs from 'fs';
import * as path from 'path';
import * as os from 'os';

const WALLET_DIR = path.join(os.homedir(), '.phettagotchi');
const KEYPAIR_FILE = 'keypair.json';

export function getWalletDir(): string {
  return WALLET_DIR;
}

export function getKeypairPath(dir?: string): string {
  return path.join(dir || WALLET_DIR, KEYPAIR_FILE);
}

/** Check if a wallet keypair exists on disk */
export function walletExists(dir?: string): boolean {
  return fs.existsSync(getKeypairPath(dir));
}

/** Generate a new Solana keypair and save to disk */
export function generateWallet(dir?: string): { publicKey: string; path: string } {
  const keypair = Keypair.generate();
  const targetDir = dir || WALLET_DIR;
  const targetPath = getKeypairPath(targetDir);

  fs.mkdirSync(targetDir, { recursive: true });
  fs.writeFileSync(targetPath, JSON.stringify(Array.from(keypair.secretKey)), { mode: 0o600 });

  return { publicKey: keypair.publicKey.toBase58(), path: targetPath };
}

/** Load keypair from disk */
export function loadKeypair(dir?: string): Keypair {
  const keypairPath = getKeypairPath(dir);
  if (!fs.existsSync(keypairPath)) {
    throw new Error(
      `No wallet found at ${keypairPath}. Call generateWallet() or ensureWallet() first.`
    );
  }
  const data = JSON.parse(fs.readFileSync(keypairPath, 'utf-8'));
  return Keypair.fromSecretKey(Uint8Array.from(data));
}

/** Get the public key (wallet address) from the stored keypair */
export function getPublicKey(dir?: string): string {
  return loadKeypair(dir).publicKey.toBase58();
}

/** Sign a legacy Transaction with the stored keypair */
export function signTransaction(tx: Transaction, dir?: string): Buffer {
  const keypair = loadKeypair(dir);
  tx.sign(keypair);
  return Buffer.from(tx.serialize());
}

/**
 * Ensure a wallet exists — generate one if it doesn't.
 * Returns the public key, file path, and whether it was newly created.
 */
export function ensureWallet(dir?: string): { publicKey: string; path: string; created: boolean } {
  if (walletExists(dir)) {
    return { publicKey: getPublicKey(dir), path: getKeypairPath(dir), created: false };
  }
  const result = generateWallet(dir);
  return { ...result, created: true };
}

/**
 * Build, sign, and return a base64-encoded signed transaction.
 * Takes the base64 unsigned tx from the build-tx API response.
 */
export function signBase64Transaction(unsignedTxBase64: string, dir?: string): string {
  const keypair = loadKeypair(dir);
  const txBytes = Buffer.from(unsignedTxBase64, 'base64');
  const tx = Transaction.from(txBytes);
  tx.sign(keypair);
  return tx.serialize().toString('base64');
}

File: packages/python/phettagotchi/wallet.py
"""Phettagotchi Wallet — Auto-generate and manage a Solana keypair.

Stores keypair at ~/.phettagotchi/keypair.json
"""

from __future__ import annotations

import json
import os
import stat
from pathlib import Path
from typing import Optional, Tuple

WALLET_DIR = Path.home() / ".phettagotchi"
KEYPAIR_FILE = "keypair.json"


def get_keypair_path(directory: Optional[Path] = None) -> Path:
    return (directory or WALLET_DIR) / KEYPAIR_FILE


def wallet_exists(directory: Optional[Path] = None) -> bool:
    return get_keypair_path(directory).exists()


def generate_wallet(directory: Optional[Path] = None) -> Tuple[str, Path]:
    """Generate a new Solana keypair and save to disk.

    Returns (public_key, keypair_path).
    Requires solders or solana-py: ``pip install solders``
    """
    try:
        from solders.keypair import Keypair  # type: ignore
    except ImportError:
        raise ImportError(
            "Install solders for wallet generation: pip install solders"
        )

    kp = Keypair()
    target_dir = directory or WALLET_DIR
    target_dir.mkdir(parents=True, exist_ok=True)
    target_path = get_keypair_path(target_dir)
    target_path.write_text(json.dumps(list(bytes(kp))))
    # Restrict permissions (owner-only read/write)
    try:
        target_path.chmod(stat.S_IRUSR | stat.S_IWUSR)
    except OSError:
        pass  # Windows may not support chmod

    return str(kp.pubkey()), target_path


def load_keypair(directory: Optional[Path] = None):
    """Load keypair from disk. Returns a solders.keypair.Keypair."""
    try:
        from solders.keypair import Keypair  # type: ignore
    except ImportError:
        raise ImportError("Install solders to load wallet: pip install solders")

    kp_path = get_keypair_path(directory)
    if not kp_path.exists():
        raise FileNotFoundError(
            f"No wallet found at {kp_path}. Call generate_wallet() first."
        )
    data = json.loads(kp_path.read_text())
    return Keypair.from_bytes(bytes(data))


def get_public_key(directory: Optional[Path] = None) -> str:
    """Get the wallet address from the stored keypair."""
    return str(load_keypair(directory).pubkey())


def ensure_wallet(directory: Optional[Path] = None) -> Tuple[str, Path, bool]:
    """Ensure a wallet exists, generating one if needed.

    Returns (public_key, keypair_path, was_created).
    """
    if wallet_exists(directory):
        return get_public_key(directory), get_keypair_path(directory), False
    pubkey, path = generate_wallet(directory)
    return pubkey, path, True


def sign_base64_transaction(unsigned_tx_b64: str, directory: Optional[Path] = None) -> str:
    """Sign a base64-encoded unsigned transaction and return base64-encoded signed tx."""
    import base64

    try:
        from solders.keypair import Keypair  # type: ignore
        from solders.transaction import Transaction  # type: ignore
    except ImportError:
        raise ImportError("Install solders for signing: pip install solders")

    kp = load_keypair(directory)
    tx_bytes = base64.b64decode(unsigned_tx_b64)
    tx = Transaction.from_bytes(tx_bytes)
    # solders Transaction signing
    signed = tx.sign([kp], tx.message.recent_blockhash)
    return base64.b64encode(bytes(signed)).decode()
